### DISCUSSION TITLE
All Micropython Lilygo T-watch OS implementation

### DISCUSSION THREAD

**Discussion by jeffmer** [micropython/micropython#10741] [Show and tell]
I bought a Lilygo T-watch V1 a few years ago and initially programmed it in Espruino. Recently, I returned to it and this time used Micropython together with a generic ESP32 SPIRAM build to program it. So unlike other T-watch projects all of the device drivers, GUI etc are programmed in Micropython. The most recent builds of Micropython with `Framebuffer.ellipse` and `Framebuffer.poly` help enormously with graphics performance. Below are some screenshots:

![](https://github.com/jeffmer/TTGO-T-watch-2020-Micropython-OS/blob/master/images/small/clockface.png?raw=true) ![](https://github.com/jeffmer/TTGO-T-watch-2020-Micropython-OS/blob/master/images/small/nifty.png?raw=true) ![](https://github.com/jeffmer/TTGO-T-watch-2020-Micropython-OS/blob/master/images/small/calculator.png?raw=true) ![]( https://github.com/jeffmer/TTGO-T-watch-2020-Micropython-OS/blob/master/images/small/weather.png?raw=true)

Battery life is about a day and half with light usage for V1 and about two and half days for V2 as V2 of the T-watch lets you hibernate and reawaken the touch device. 

For anyone interested, full source and some documentation at https://github.com/jeffmer/TTGO-T-watch-2020-Micropython-OS

**Comment by andypiper** [comment]
This is cool, thanks for sharing. I have the V3 I think, so I'll have to have a look at the differences.

(unrelated - if you're near that location, I'm only a stone's throw away!)

**Reply by jeffmer** [reply]
Thanks,

From the documentation V3 looks like V1 for pinout and buzzer motor with the addition of a touch controller reset GPIO pin. Also, accelerometer orientation may be different. 

( yes from your profile, it does look like we are close!)

**Comment by varna9000** [comment]
Brilliant! Had to dig out my v1 watch when I saw this. I'll defo have to make a module for it.

**Comment by jeffmer** [comment]
Thanks, I have now had to resort to my own build of the firmware to support the full 16Mb of flash and GPIO wake-up from light sleep. If you do try it, happy to help if you have any issues.

**Reply by karfas** [reply]
This is a great example implementation, thank you very much for sharing.

I compiled the uPy image myself, freezing some of your modules (where possible without changes).

Had to restart the watch 2-3 times since yesterday and will most likely add some logging to find a cause - the "lockups" happened during use of the touch screen, I think.

Thank you very much again!

**Reply by jeffmer** [reply]
I have been working on a V2 recently and I may have introduced some bugs for a V1.  Also, if you built your own image,  did you find my GPIO wake-up mods here https://github.com/jeffmer/micropython/blob/GPIO_WAKEUP/ports/esp32/machine_pin.c?

**Reply by karfas** [reply]
Of course - the AXP driver raises an exception without the patch. 

Even when I would have preferred to use some stock micropython, I merged the GPIO_WAKEUP branch from your micropython fork (some hint/link in the t-watch README would have been useful). 
I am not convinced that this is the best way to handle the wakeup from light sleep, but will investigate and give you feedback in the other thread (in a week or so).

**Reply by jeffmer** [reply]
Feedback would be very much appreciated. I will include a link in the README as you suggest. I would have preferred to use a standard  build but I could not see a way of waking up the watch from multiple sources without either polling by waking up the watch every second to test or the patch which is better for battery life. BTW. If for some reason the touch controller goes into hibernate mode there is no way if resetting it on V1 watches other than powering off and on again - would that explain your lockups?

**Reply by karfas** [reply]
Not sure if these are really lockups. I had removed the applications (they are great, but not useful for me), so a swipe up should do nothing.
However, the second hand on the watch just freezes after the swipe up; a swipe down in this state "activates" the watch again. 
Most likely, I have misinterpreted this behaviour (will observe this further during the next days).

**Comment by jeffmer** [comment]
Ah, then most likely when you swipe up there is no app to run so the screen will not be redrawn and you will be left with the clock image. When you swipe down, the clock will start running again.

**Reply by jeffmer** [reply]
With the latest update, by reducing supply voltage from 3.3V to 2.7V during sleep, the V1 light sleep current is now 2.0ma and the V2 is 1.3ma so battery life should be increased by about 50%. The bad news is that there is a residual current of 0.3ma when the watch is turned off - presumably consumed by the axp202 power management chip - so the battery will go completely flat in a month or so even if the watch is turned off.
